Add options for loading tiles via AJAX and custom AJAX request headers

jonathangreen opened the pull request:
## JIRA Ticket

Part of: 
[ISLANDORA-2100](https://jira.duraspace.org/browse/ISLANDORA-2100)

Upstream PR here: 
https://github.com/internetarchive/bookreader/pull/59

I know we haven't had much luck in the past getting things merged upstream, but I wanted to try.

## What does this Pull Request do?

This PR is based in spirit on https://github.com/openseadragon/openseadragon/pull/1055. It adds the option to load book pages using AJAX and sending custom request headers. This is useful if you have book pages that need authentication to be served that can be sent with the headers. This feature is very useful when using token-based authentication.

When the option not set then tiles should load as they did before.

## What's new?

### property: `loadWithAjax` (Boolean)
When this option is set on the `BookReader` all of the images are loaded in with AJAX.

### function: `getAjaxHeaders(index)`
When this function is provided on the `BookReader` it will be called with the index of the page and returns a  `String -> String` mapping of headers to be sent with the book page. This allows us to send different headers for every page if desired.

## How should this be tested?

For code review, be aware that there are 3 commits here. Two are just code style cleanup. One cleans the whitespace at the end of lines and the other fixes some spelling errors in the source. These have already been cleaned up upstream. The real meat of this PR is https://github.com/jonathangreen/internet_archive_bookreader/commit/cb61e49051ab904788ac19676826d2dca7510f85.

There is a forthcoming PR to the [Islandora IA Bookreader](https://github.com/Islandora/islandora_internet_archive_bookreader) module as part of [ISLANDORA-2100](https://jira.duraspace.org/browse/ISLANDORA-2100) that will enable this option though the UI for testing.

It can also be tested by manually setting the options as described above.

* Make sure that books continue to load as before with `loadWithAjax` set to `false`.
* Make sure books load properly with `loadWithAjax` set to `true`.
* Make sure loads send the correct headers when using `loadWithAjax` with `ajaxHeaders` set.

## Additional Notes:

Setting the `loadWithAjax` option will limit browser support to IE 10+ due to the use of responseType = 'arraybuffer' and [createObjectURL](http://caniuse.com/#feat=bloburls). However if this option isn't used, then the behaviour is the same as it was before. 

## Interested parties
@qadan @Islandora/7-x-1-x-committers

jonathangreen commented:
@whikloj did you test this at all when you were testing https://github.com/Islandora/islandora_internet_archive_bookreader/pull/113?

whikloj commented:
@jonathangreen I did not test this, I'll have to have a look and see if I can figure this out.

jonathangreen commented:
@whikloj no worries. Just wanted to see if you had already tested it. I'm going to bump for @Islandora/7-x-1-x-committers too, in case anyone feels like digging into some JS. 

The bulk of the changes are whitespace, so its not actually as big a change as it seems, https://github.com/Islandora/internet_archive_bookreader/pull/8/commits/cb61e49051ab904788ac19676826d2dca7510f85 contains just the code changes. Might be easier to review that to take a look at the code.

DiegoPino commented:
@jonathangreen this looks good. Wondering if you could actually only pull https://github.com/Islandora/internet_archive_bookreader/commit/cb61e49051ab904788ac19676826d2dca7510f85 instead of the one with all the space changes? I have a pull pending that uses IABV2 and it would make my life much easier if i could port your changes to that. Finally. Is there anything of this that could be ported to our[ extended .js version ](https://github.com/Islandora/islandora_internet_archive_bookreader/blob/7.x/js/islandora_book_reader.js) instead of here? I mean, yes, it would duplicate code by overriding a lot of functions but would give us certain independence and certain portability.  Not sure if that makes sense. thanks a lot.

DiegoPino commented:
And sorry for the delay, i don't even sleep lately... 😴

jonathangreen commented:
Hi @DiegoPino. Sorry about the delay responding and thanks for the review. I've also been very busy, but will try to address your concerns soon.

jonathangreen commented:
@DiegoPino I've updated the PR to only pull code changes. When you are updating to IABV v2 you can apply the patch here https://github.com/internetarchive/bookreader/pull/59 which does the exact same thing as this, just rebased onto IABVv2. 

I could move this code over to [our extended version](https://github.com/Islandora/islandora_internet_archive_bookreader/blob/7.x/js/islandora_book_reader.js). I would need to totally override: [`BookReader.prototype.prefetchImg`](https://github.com/jonathangreen/internet_archive_bookreader/blob/7a407c3acaab695500e89b2152c972ef543b058b/BookReader/BookReader.js#L2533-L2561), [`BookReader.prototype.drawLeafsThumbnail`](https://github.com/jonathangreen/internet_archive_bookreader/blob/7a407c3acaab695500e89b2152c972ef543b058b/BookReader/BookReader.js#L594-L823), [`BookReader.prototype.lazyLoadImage`](https://github.com/jonathangreen/internet_archive_bookreader/blob/7a407c3acaab695500e89b2152c972ef543b058b/BookReader/BookReader.js#L848-L887) and add [`BookReader.prototype.loadImage`](https://github.com/jonathangreen/internet_archive_bookreader/blob/7a407c3acaab695500e89b2152c972ef543b058b/BookReader/BookReader.js#L412-L464). This would duplicate ~300 lines of the book reader code in our own JS file. Is it worth it to duplicate that much code? To me it seems like this approach just makes pulling in upstream changes harder, as there is no git history to follow and merge, all the merging has to be done by hand. 

If there is consensus that duplicating the code in our own book reader override is better though, I can make that change, but its not my preferred option.

DiegoPino commented:
@jonathangreen do you think you can add the UI commit soon to enable this option to be togglable for this release?  Merging now, in case something does not work under certain circumstances I should be blamed, because testing JS in every browser/situation is hard.
